base/subarray.jl

```diff
@@ -461,3 +461,38 @@ function _indices_sub(i1::AbstractArray, I...)
 end
 
 has_offset_axes(S::SubArray) = has_offset_axes(S.indices...)
+
+# faster implementation of folding operations for "reversed" FastSubArray with stride1 == -1
+
+function reverse_fastsubarray_1(a)
+# here we assume a.stride1 == -1
+    fi, li = a.offset1 .- (firstindex(a), lastindex(a))
+    view(parent(a), li:fi)
+end
+
+function mapfoldl_impl(f::F, op::OP, init::I, a::FastSubArray) where {F,OP,I}
+    if a.stride1 == -1
+        b = reverse_fastsubarray_1(a)
+        @invoke mapfoldr_impl(f::F, FlipArgs(op)::FlipArgs{OP}, init::I, b::AbstractArray)
+    else
+        @invoke mapfoldl_impl(f::F, op::OP, init::I, a::AbstractArray)
+    end
+end
+
+function mapfoldr_impl(f::F, op::OP, init::I, a::FastSubArray) where {F,OP,I}
+    if a.stride1 == -1
+        b = reverse_fastsubarray_1(a)
+        @invoke mapfoldl_impl(f::F, FlipArgs(op)::FlipArgs{OP}, init::I, b::AbstractArray)
+    else
+        @invoke mapfoldr_impl(f::F, op::OP, init::I, a::AbstractArray)
+    end
+end
+
+function _mapreduce_dim(f::F, op::OP, nt::NT, a::FastSubArray, dims::Colon) where {F,OP,NT}
+    if a.stride1 == -1
+        b = reverse_fastsubarray_1(a)
+        @invoke _mapreduce_dim(f::F, FlipArgs(op)::FlipArgs{OP}, nt::NT, b::AbstractArray, dims::Colon)
+    else
+        @invoke _mapreduce_dim(f::F, op::OP, nt::NT, a::AbstractArray, dims::Colon)
+    end
+end
```

Comment:
I think we'd better overloading `_mapreduce(f, op, ::IndexLinear, ::FastSubArray)` instead.

Reply:
That would slow `mapreduce` with given `init` down because `_mapreduce` only covers the case of `mapreduce` without `init`. If `init` is given, then `mapreduce` calls `mapfoldl_impl`. For a "reversed" view, `mapfoldl_impl` is translated into `mapfoldr_impl` of the parent array. The current PR instead uses `mapfoldl_impl` of the parent array, which is 5 to 75 times faster, see the difference between `foldl` and `foldr` in the benchmarks above.

Reply:
But we have add `mapfoldl(r)_impl ` specializaiton for `FastSubArray` above?

Reply:
Yes, that's exactly the point: For `w = view(v, 3:-1:1)` the following happens with your suggestion:
```
mapreduce(f, op, w; init = i) =
mapfoldl_impl(f, op, i, w) =
mapfoldr_impl(f, FlipArgs(op), i, view(v, 1:3))
```
For such a view (with known step value 1) `mapfoldr` is much slower than `mapfoldl`. It's the same as for the underlying array:
```
julia> v = rand(Int8, 2^16); w = @view v[begin:end];
julia> @btime foldl(+, $v);
  578.033 ns (0 allocations: 0 bytes)
julia> @btime foldl(+, $w);
  644.651 ns (0 allocations: 0 bytes)
julia> @btime foldr(+, $v);
  34.613 μs (0 allocations: 0 bytes)
julia> @btime foldr(+, $w);
  39.116 μs (0 allocations: 0 bytes)
```